packages/core/integration-tests/test/lazy-compile.js

```diff
@@ -0,0 +1,123 @@
+import assert from 'assert';
+import path from 'path';
+import {
+  bundler,
+  outputFS,
+  distDir,
+  getNextBuild,
+  assertBundles,
+  removeDistDirectory,
+  run,
+} from '@parcel/test-utils';
+
+const findBundle = (bundleGraph, nameRegex) => {
+  return bundleGraph.getBundles().find(b => nameRegex.test(b.name));
+};
+
+const distDirIncludes = async matches => {
+  const files = await outputFS.readdir(distDir);
+  for (const match of matches) {
+    if (typeof match === 'string') {
+      if (!files.some(file => file === match)) {
+        throw new Error(
+          `No file matching ${match} was found in ${files.join(', ')}`,
+        );
+      }
+    } else {
+      if (!files.some(file => match.test(file))) {
+        throw new Error(
+          `No file matching ${match} was found in ${files.join(', ')}`,
+        );
+      }
+    }
+  }
+  return true;
+};
+
+describe('lazy compile', function () {
+  it('should lazy compile', async function () {
+    const b = await bundler(
+      path.join(__dirname, '/integration/lazy-compile/index.js'),
+      {
+        shouldBuildLazily: true,
+        mode: 'development',
+        shouldContentHash: false,
+      },
+    );
+
+    await removeDistDirectory();
+
+    const subscription = await b.watch();
+    let result = await getNextBuild(b);
+
+    // This simulates what happens if index.js is loaded as well as lazy-1 which loads lazy-2.
+    // While parallel-lazy-1 is also async imported by index.js, we pretend it wasn't requested (i.e. like
+    // if it was behind a different trigger).
+    result = await result.requestBundle(
+      findBundle(result.bundleGraph, /index.js/),
+    );
+    result = await result.requestBundle(
+      findBundle(result.bundleGraph, /^lazy-1/),
+    );
+    result = await result.requestBundle(
+      findBundle(result.bundleGraph, /^lazy-2/),
+    );
+
+    // Expect the bundle graph to contain the whole nest of lazy from `lazy-1`, but not
+    // `parallel-lazy-1` which wasn't requested.
+    assertBundles(result.bundleGraph, [
+      {
+        assets: ['index.js', 'bundle-url.js', 'cacheLoader.js', 'js-loader.js'],
+      },
+      {
+        assets: ['lazy-1.js', 'esmodule-helpers.js'],
+      },
+      {
+        assets: ['lazy-2.js'],
+      },
+      {
+        assets: ['parallel-lazy-1.js'],
+      },
+    ]);
+
+    subscription.unsubscribe();
+
+    // Ensure the files match the bundle graph - lazy-2 should've been produced as it was requested
+    assert(await distDirIncludes(['index.js', /^lazy-1\./, /^lazy-2\./]));
+  });
+
+  it('should lazy compile properly when same module is used sync/async', async () => {
```

Comment:
This test still fails on this branch - the "fix" is to remove the conditions for `shouldBuildLazily` in `JSRuntime` (e.g.: https://github.com/parcel-bundler/parcel/blob/7ff54c7c1212b05ed0393e0a92e96c6002d1dc51/packages/runtimes/js/src/JSRuntime.js#L366)

With the code as it is, the whole bundle group doesn't get loaded - so when the `uses-static-component-async` bundle loads, it doesn't trigger a load of the `static-component` bundle, so it can't resolve the module.

I know that "fix" (seen here: https://github.com/marcins/parcel/commit/a36a37d8add87b90d63fa941b759f4c3baf9dadb) means that there is a potential issue when a bundle group grows after lazy compilation - this means Parcel can get into a state where it requires a browser refresh in order to recover. However, with the current state, a refresh won't heal as the load order is just incorrect.

Reply:
As discussed in the Core meeting, the "fix" for this has now been cherry-picked onto this branch. I've updated the comments to leave behind some trace that there was lazy build specific code previously there.